File: game/src/main/java/net/demilich/metastone/game/behaviour/sarsa_mlp.py
# -*- coding: utf-8 -*-
"""
Created on Thu Jul 13 11:27:38 2017

@author: sjxn2423

可以尝试用Sarsa的方式构造数据 Sutton P230

// 尝试使用Sarsa方法训练一个基于MLP的value function （sarsa本来是在线更新的，现在我们尝试一局batch更新一次，不确定理论上是否靠谱，后面可以尝试标准的在线更新Linear模型）
// 训练的loss函数： L(w) = [r_t + gamma*Q(s_t+1, a_t+1, w_fix) - Q(s_t, a_t, w)]^2
// 将执行action之后到达的局面s_t+1的描述向量作为 (s_t, a_t) 的描述, 训练目标变成 [r_t + gamma*Q(s_t+2, w_fix)} - Q(s_t+1, w)]^2
// 暂时考虑将我方的每一次action算作一个time step，我方EndTurn动作和对方完整执行以及下一个turn己方发牌合在一起算作一个time step （将对方当做环境），需要记录每一次操作的数据

// 后面也可以尝试将我方一个turn内的完整action path当做一个action，以一个完整回合作为一个time step，似乎更合理，因为我们更希望评估的是一系列操作之后的状态优劣
// sarsa似乎可以很容易改成那种action path定义，因为不需要像Q-Learning那样需要计算argmax，而直接follow epsi-greedy 策略即可

"""

import os
import cPickle as pickle
import random
from collections import defaultdict
from sklearn.neural_network import MLPRegressor
#from sklearn.externals import joblib
from sklearn2pmml import PMMLPipeline
from sklearn2pmml import sklearn2pmml

data_file = 'E:\\MetaStone\\app\\report.log'
model_file = 'E:\\MetaStone\\app\\mlp.model'
pmml_file = 'E:\\MetaStone\\app\\mlp.pmml'


data_buffer = []

if os.path.exists(model_file):
#    mlp = joblib.load(model_file)
    mlp = pickle.load(open(model_file, 'rb'))
    mlp.set_params(learning_rate_init=0.0002)  #尝试减小更新步子
else:
    mlp = MLPRegressor(hidden_layer_sizes=(5,), learning_rate_init=2e-4, tol=1e-6, random_state=3)
      
def get_immediate_reward(state_next, state):
    """根据连续状态下Hp的相对变化计算reward"""
    r = state_next[0]-state_next[15] - (state[0]-state[15])
    return r

def get_sarsa_data(data_dict, final_hp_diff, total_turn, final_env_state):
    """根据sarsa算法，提取完整一局的特征数据和对应标签, 
    immediate reward使用执行action之后带来的双方Hp变化量"""
    gamma = 0.99
    for turn in range(1, total_turn-1):
        reward = get_immediate_reward(data_dict[turn]['envState'], data_dict[turn-1]['envState'])
        target = reward + gamma*data_dict[turn+1]['QScore']
        data_buffer.append((data_dict[turn]['envState'], target))        
    data_buffer.append((data_dict[total_turn-1]['envState'], final_hp_diff))        
                         
def load_data(file_name):
    data_dict = defaultdict(dict)
    with open(file_name, 'r') as f:
        for line in f:
            if line[0] != '{':
                continue            
            info = eval(line)            
            if 'winner' in info: #一局结束, 提出完整一局的特征数据和对应标签
                get_sarsa_data(data_dict, info['HpDiff'], info['Turn'], info['envState'])
            else:
                data_dict[info['Turn']]['envState'] = info['envState']
                data_dict[info['Turn']]['QScore'] = info['QScore']         

if __name__ == '__main__':
    
    load_data(data_file)
    # 清空report.log
    with open(data_file, 'w') as f:
        f.write('')
    
    X_batch = [data[0] for data in data_buffer]
    y_batch = [data[1] for data in data_buffer]
    
    # partial_fit是incremental更新，看了partial_fit的源代码，它更新一步之后就break了，所以和max_iter无关，它主要受learning_rate_init的数值影响，就是步子的大小
    mlp.partial_fit(X_batch, y_batch)
#    print mlp.intercepts_
#    mlp.predict(X_batch[:5])
    
    print "number of samples: {}".format(len(y_batch))
    pickle.dump(mlp, open(model_file, 'wb'))     

    # dump the model to PMML file
    mlp_pipeline = PMMLPipeline([("Q_model", mlp)])
    sklearn2pmml(mlp_pipeline, pmml_file, with_repr = True)

File: game/src/main/java/net/demilich/metastone/game/behaviour/SarsaMLP.java
package net.demilich.metastone.game.behaviour;

import net.demilich.metastone.game.GameContext;
import net.demilich.metastone.game.Player;
import net.demilich.metastone.game.actions.GameAction;
import net.demilich.metastone.game.behaviour.heuristic.IGameStateHeuristic;
import net.demilich.metastone.game.cards.Card;
import net.demilich.metastone.game.logic.GameLogic;
import org.dmg.pmml.FieldName;
import org.dmg.pmml.PMML;
import org.jpmml.evaluator.*;
import org.jpmml.model.PMMLUtil;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.*;
import java.util.*;

// 尝试使用Sarsa方法训练一个基于MLP的value function （sarsa本来是在线更新的，现在我们尝试一局batch更新一次，不确定理论上是否靠谱，后面可以尝试标准的在线更新Linear模型）
// 训练的loss函数： L(w) = [r_t + gamma*Q(s_t+1, a_t+1, w_fix) - Q(s_t, a_t, w)]^2
// 将执行action之后到达的局面s_t+1的描述向量作为 (s_t, a_t) 的描述, 训练目标变成 [r_t + gamma*Q(s_t+2, w_fix)} - Q(s_t+1, w)]^2
// 暂时考虑将我方的每一次action算作一个time step，我方EndTurn动作和对方完整执行以及下一个turn己方发牌合在一起算作一个time step （将对方当做环境），需要记录每一次操作的数据

// 后面也可以尝试将我方一个turn内的完整action path当做一个action，以一个完整回合作为一个time step，似乎更合理，因为我们更希望评估的是一系列操作之后的状态优劣
// sarsa似乎可以很容易改成那种action path定义，因为不需要像Q-Learning那样需要计算argmax，而直接follow epsi-greedy 策略即可

// 注意：每回合开始时双方随机抽牌也会对游戏局面引入随机干扰

public class SarsaMLP extends Behaviour {

	private final static Logger logger = LoggerFactory.getLogger(SarsaMLP.class);
	private Random random = new Random();
	private double epsilon = 0.1;
	private int turn = 0;
	private final static String pmmlFile = "E:\\MetaStone\\app\\mlp.pmml";
	private static Evaluator modelEvaluator;
	private long[] playersWinCount = {0,0};
	private static double runningReward = 0;

	public SarsaMLP() {
		File file = new File(pmmlFile);
		//如果PMML文件存在，直接加载已有模型
		if (file.exists()) {
			try {
				loadPMMLModel(pmmlFile);
			} catch (Exception e) {
				e.printStackTrace();
			}
		} else{
			this.modelEvaluator = null;
		}
	}

	@Override
	public String getName() {
		return "Sarsa-MLP";
	}

	@Override
	public List<Card> mulligan(GameContext context, Player player, List<Card> cards) {
		List<Card> discardedCards = new ArrayList<Card>();
		for (Card card : cards) {
			if (card.getBaseManaCost() >= 4) {  //耗法值>=4的不要
				discardedCards.add(card);
			}
		}
		return discardedCards;
	}

	@Override
	public GameAction requestAction(GameContext context, Player player, List<GameAction> validActions) {

		// 记录需要的数据
		Player opponent = context.getOpponent(player);
		List<Integer> envState = player.getPlayerState();
		envState.addAll(opponent.getPlayerState());
		double QScore = evaluateContext(context, player.getId());
		logger.info("{'Turn': {}, 'envState': {}, 'QScore': {}}", this.turn, envState.toString(), QScore);
		this.turn += 1;

		// get epsilon greedy action
		if (random.nextDouble() < epsilon) {  //返回随机action
			int randomIndex = random.nextInt(validActions.size());
			GameAction randomAction = validActions.get(randomIndex);
			return randomAction;
		} else {
			// get best action at the current state and the corresponding Q-score
			GameAction bestAction = validActions.get(0);
			double bestScore = Double.NEGATIVE_INFINITY;
			for (GameAction gameAction : validActions) {
				GameContext simulationResult = simulateAction(context.clone(), player, gameAction);  //假设执行gameAction，得到之后的game context
				double gameStateScore = evaluateContext(simulationResult, player.getId());   //heuristic评估执行gameAction之后的游戏局面的分数
				if (gameStateScore > bestScore) {		// 记录得分最高的action
					bestScore = gameStateScore;
					bestAction = gameAction;
				}
				simulationResult.dispose();  //GameContext环境每次仿真完销毁
			}
			return bestAction;  // 返回argmax_a{Q(s_t,a)
		}
	}

	private GameContext simulateAction(GameContext simulation, Player player, GameAction action) {
		GameLogic.logger.debug("");
		GameLogic.logger.debug("********SIMULATION starts********** " + simulation.getLogic().hashCode());
		simulation.getLogic().performGameAction(player.getId(), action);   // 在simulation GameContext中执行action，似乎是获取logic模块来执行action的
		GameLogic.logger.debug("********SIMULATION ends**********");
		GameLogic.logger.debug("");
		return simulation;
	}

	private void loadPMMLModel(String file) throws Exception {
		try(InputStream model = new FileInputStream(new File(file))){
			PMML pmml = PMMLUtil.unmarshal(model);
			ModelEvaluatorFactory modelEvaluatorFactory = ModelEvaluatorFactory.newInstance();
			this.modelEvaluator = modelEvaluatorFactory.newModelEvaluator(pmml);
//			logger.info("PMML evaluator updated!");
//			logger.info("PMML is null: {}", this.modelEvaluator==null);
//			logger.info("PMML targetFields {}", this.modelEvaluator.getTargetFields());
		} catch(Exception e) {
			logger.error(e.toString());
			throw e;
		}
	}

	private double evaluateContext(GameContext context, int playerId) {
		if (this.modelEvaluator==null){
			logger.info("PMML evaluator is null");
			return random.nextDouble();
		}
		Player player = context.getPlayer(playerId);
		Player opponent = context.getOpponent(player);
//		if (player.getHero().isDestroyed()) {   // 己方被干掉，得分 负无穷
//			return Float.NEGATIVE_INFINITY;  // 正负无穷会影响envState的解析，如果要加的话可以改成 +-100之类的
//		}
//		if (opponent.getHero().isDestroyed()) {  // 对方被干掉，得分 正无穷
//			return Float.POSITIVE_INFINITY;
//		}
		List<Integer> envState = player.getPlayerState();
		envState.addAll(opponent.getPlayerState());

//		logger.info("envState:{}", envState);
		// 准备模型输入数据
		Map<FieldName, FieldValue> arguments = new LinkedHashMap<>();
		List<InputField> inputFields = this.modelEvaluator.getInputFields();
		int i = 0;
		for(InputField inputField : inputFields){
			FieldName inputFieldName = inputField.getName();
			Object rawValue = envState.get(i);
			FieldValue inputFieldValue = inputField.prepare(rawValue);
			arguments.put(inputFieldName, inputFieldValue);
			i++;
		}
		//	logger.info("PMML inputs {}", arguments);
		Map<FieldName, ?> results = this.modelEvaluator.evaluate(arguments);
		List<TargetField> targetFields = this.modelEvaluator.getTargetFields();
		//	logger.info("PMML results {}", results);
		//	logger.info("PMML targetFields {}", targetFields);
		Object targetFieldValue = results.get(targetFields.get(0).getName());
//		logger.info("PMML targetFieldValue {}", targetFieldValue);
//		return envState.get(0)-envState.get(15);
		return (double)targetFieldValue;
	}

	@Override
	public void onGameOver(GameContext context, int playerId, int winningPlayerId) {
		// GameOver的时候会跳入这个函数
		if(winningPlayerId <= 1)
			playersWinCount[winningPlayerId] += 1;

		Player player = context.getPlayer(playerId);
		Player opponent = context.getOpponent(player);
		List<Integer> envState = player.getPlayerState();
		envState.addAll(opponent.getPlayerState());
		int HpDiff =  player.getHero().getHp() - opponent.getHero().getHp();
		logger.info("{'Turn': {}, 'envState': {}, 'winner': {}, 'HpDiff': {}}", this.turn, envState.toString(), winningPlayerId, HpDiff);  // 可以根据HpDiff来判断胜负和设定reward
		this.turn = 0;
		if (this.runningReward == 0){
			this.runningReward = (double)HpDiff;
		}else{
			this.runningReward = this.runningReward*0.95 + HpDiff*0.05;
		}
		logger.info("##################Winning Player is : {} ##################", winningPlayerId);
		logger.info("playersWinCount: {}, player0 win ratio: {}, running reward: {}", playersWinCount, 1.0*playersWinCount[0]/(playersWinCount[0]+playersWinCount[1]), runningReward);

		try {
			// 记录胜率情况到文件
			FileWriter fileWriter = new FileWriter(new File("E:\\MetaStone\\app\\winStat.log"), true);
			String s = String.format("player0 wins:%d, player1 wins:%d, player0 win ratio:%f, running reward:%f\n",
					playersWinCount[0],playersWinCount[1],1.0*playersWinCount[0]/(playersWinCount[0]+playersWinCount[1]),runningReward);
			fileWriter.write(s);
			fileWriter.close();

			// 使用最新一局的数据更新模型，调用外部Python脚本，保存更新后的模型到PMML文件
			Process pr = Runtime.getRuntime().exec("D:\\Anaconda\\python.exe E:\\MetaStone\\game\\src\\main\\java\\net\\demilich\\metastone\\game\\behaviour\\sarsa_mlp.py");
			BufferedReader br = new BufferedReader(new InputStreamReader(pr.getInputStream()));
			String line;
			while ((line = br.readLine()) != null) {
				logger.info(line);
			}
			br.close();
			int exitValue = pr.waitFor();
			if (exitValue!=0)
				logger.info("exitValue: " + exitValue + "Attention: There must be something wrong in the Python script!!!!!!!!!!!!!!!!!!");  // exitValue = 0 表示python子进程正常退出，如果为1可能是Python代码报错了

			// 加载python训练保存的PMML文件
			loadPMMLModel(pmmlFile);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}
}
